src/process.c

```diff
@@ -463,59 +466,450 @@ cc_run_hook(struct oci_cfg_hook* hook, const gchar* state,
+ * config.
+ *
+ * \param config \ref cc_oci_config.
+ * \param netcfg Newline-separated "name=value" format
+ *   network configuration.
+ *
+ * \return \c true on success, else \c false.
+ */
+static gboolean
+cc_oci_vm_netcfg_from_str (struct cc_oci_config *config,
+		const gchar *netcfg)
+{
+	gchar  **lines = NULL;
+	gchar  **line;
+
+	if (! (config && netcfg)) {
+		return false;
+	}
+
+	lines = g_strsplit (netcfg, "\n", -1);
+	if (! lines) {
+		return false;
+	}
+
+	for (line = lines; line && *line; line++) {
+		gchar **fields;
+
+		fields = g_strsplit (*line, "=", 2);
+		if (! fields) {
+			break;
+		}
+
+		if (! g_strcmp0 (fields[0], "@HOSTNAME@")) {
+			config->net.hostname = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@GATEWAY@")) {
+			config->net.gateway = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@DNS_IP1@")) {
+			config->net.dns_ip1 = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@DNS_IP2@")) {
+			config->net.dns_ip2 = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@MAC_ADDRESS@")) {
+			config->net.mac_address = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@IPV6_ADDRESS@")) {
+			config->net.ipv6_address = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@IP_ADDRESS@")) {
+			config->net.ip_address = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@SUBNET_MASK@")) {
+			config->net.subnet_mask = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@IFNAME@")) {
+			config->net.ifname = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@BRIDGE@")) {
+			config->net.bridge = g_strdup (fields[1]);
+		} else if (! g_strcmp0 (fields[0], "@TAP_DEVICE@")) {
+			config->net.tap_device = g_strdup (fields[1]);
+		}
+
+		g_strfreev (fields);
+	}
+
+	g_strfreev (lines);
+
+	return true;
+}
+
 /*!
  * Start the hypervisor (in a paused state) as a child process.
  *
+ * Due to the way networking is handled in Docker, the logic here
+ * is unfortunately rather complex.
+ *
  * \param config \ref cc_oci_config.
  *
  * \return \c true on success, else \c false.
  */
 gboolean
 cc_oci_vm_launch (struct cc_oci_config *config)
 {
-	gchar     **args = NULL;
-	GError     *err = NULL;
-	gboolean    ret;
-	GPid        pid;
-	int         status = 0;
-	GSpawnFlags flags = 0x0;
-	gchar     **p;
+	gboolean           ret = false;
+	GPid               pid;
+	int                status = 0;
+	size_t             len = 0;
+	ssize_t            bytes;
+	char               buffer[2];
+	int                net_cfg_pipe[2] = {-1, -1};
+	int                child_err_pipe[2] = {-1, -1};
+	gchar            **args = NULL;
+	gchar            **p;
+	gchar            **netcfgv = NULL;
+	g_autofree gchar  *final = NULL;
+	g_autofree gchar  *timestamp = NULL;
+
+	timestamp = cc_oci_get_iso8601_timestamp ();
+	if (! timestamp) {
+		goto out;
+	}
+
+	config->state.status = OCI_STATUS_CREATED;
```

Comment:
you are changing the status to created before create the VM, what will happen if the VM fails ? will we have an inconsistent state?